### PR TITLE
Add "disable taps during gameplay" to touch input settings and related UI

### LINKED ISSUE #25065: Touchscreen "Click/Tap Streaming" No Longer Working
### Type

Game behaviour

### Bug description

Touchscreen functions including: cursor, scrolling, etc. all are recorded. However, "click" inputs such as tapping on notes no longer work. Osu! Mania/other key rhythm games function fine. Just standard osu! does not recognize tap/click inputs.

Updated to latest version of osu!lazer via automatic updates on Testflight on IOS 16.3.1

Says there is a newer version (2023.1008.0 -> 2023.1008.1)

iPhone SE (2020), both fingers and stylus do not function for input. Both, however, work for movement.

### Screenshots or videos

https://github.com/ppy/osu/assets/69877475/70aa9e1b-7f0c-40b5-af7c-ea38ca2c924b

### Version

2023.1008.0-lazer

### Logs

[database.log](https://github.com/ppy/osu/files/12843017/database.log)
[input.log](https://github.com/ppy/osu/files/12843018/input.log)
[network.log](https://github.com/ppy/osu/files/12843019/network.log)
[performance.log](https://github.com/ppy/osu/files/12843020/performance.log)
[runtime.log](https://github.com/ppy/osu/files/12843021/runtime.log)

### PR DESCRIPTION
**Pull request by Susko3**
- Closes https://github.com/ppy/osu/issues/25065

Adds a new setting (`OsuSetting.GameplayDisableTaps`) and changes existing strings to be in line with https://github.com/ppy/osu/issues/25065#issuecomment-1786360644.

| Area | Old string | New string | Notes |
|--------|--------|--------|--------|
| `TouchSettings` for all platforms | N/A | "Disable taps during gameplay" | Newly added |
| beatmap loading screen `InputSettings` | "Disable mouse buttons during gameplay"  | "Disable clicks during gameplay" or "Disable taps during gameplay" | currently based on `RuntimeInfo.IsDesktop`, but can be changed to use https://github.com/ppy/osu/pull/25348|
| SDL2 and Android `MouseSettings` | "Disable mouse buttons during gameplay" | "Disable clicks during gameplay" | |

### CODE REVIEW

**Comment by peppy** [comment]
I don't understand. Should this not be adding a new tracked setting? It feels so wrong to have both toggles affect the same underlying setting. What's the point?

**Comment by Susko3** [comment]
> It feels so wrong to have both toggles affect the same underlying setting. What's the point?

I thought we were going for option 1 in https://github.com/ppy/osu/issues/25065#issuecomment-1784593467. But I can agree with adding a new setting that affects only touch (and changing the mouse setting to only affect mouse). (I haven't considered adding a new underlying setting up until now.)

> Should this not be adding a new tracked setting?

For now, I'm against adding a new hotkey (and a tracked settings to go with it).

**Comment by peppy** [comment]
By tracked setting I just meant a new `OsuConfigManager` entry. I agree it shouldn't be OSD tracked.

File: osu.Game/Rulesets/UI/RulesetInputManager.cs [hunk 1]
```diff
@@ -147,9 +149,9 @@ protected override bool Handle(UIEvent e)
 
         protected override bool HandleMouseTouchStateChange(TouchStateChangeEvent e)
         {
-            if (mouseDisabled.Value)
+            if (tapsDisabled.Value)
             {
-                // Only propagate positional data when mouse buttons are disabled.
+                // Only propagate positional data when taps are disabled.
                 e = new TouchStateChangeEvent(e.State, e.Input, e.Touch, false, e.LastPosition);
             }
```
Comment (Susko3) [inline osu.Game/Rulesets/UI/RulesetInputManager.cs:156]: This isn't doing much. All the rulesets now support touch natively and don't need mouse mouse emulation. And therefore completely ignore this input path. Possibly still relevant for custom rulesets.

File: osu.Game/Configuration/OsuConfigManager.cs [hunk 2]
```diff
@@ -412,6 +414,7 @@ public enum OsuSetting
         EditorLimitedDistanceSnap,
         ReplaySettingsOverlay,
         AutomaticallyDownloadMissingBeatmaps,
-        EditorShowSpeedChanges
+        EditorShowSpeedChanges,
+        GameplayDisableTaps,
```
Comment (frenzibyte) [inline osu.Game/Configuration/OsuConfigManager.cs:418]: I would want to include `Touch` anywhere in this setting name so I can easily find it, "Gameplay" sounds like a broad term to me for what is meant to be a touch-specific setting.

File: osu.Game/Rulesets/UI/RulesetInputManager.cs (hunk 1 above)
Comment (frenzibyte) [reply osu.Game/Rulesets/UI/RulesetInputManager.cs:156]: This path is specific to the osu! ruleset, and it should be removed now that osu! handles touch input appropriately with the `OsuTouchInputMapper` component (as well as correctly ignoring taps while still maintaining position, without doing it weirdly like in the code highlighted here).

But that should be done in a separate PR to keep review complexity to a minimum.

File: osu.Game/Overlays/Settings/Sections/Input/TouchSettings.cs [hunk 3]
```diff
@@ -3,38 +3,48 @@
...
         {
             this.handler = handler;
         }
 
         [BackgroundDependencyLoader]
-        private void load()
+        private void load(OsuConfigManager osuConfig)
         {
-            Children = new Drawable[]
+            if (!RuntimeInfo.IsMobile) // don't allow disabling the only input method (touch) on mobile.
```
Comment (frenzibyte) [inline osu.Game/Overlays/Settings/Sections/Input/TouchSettings.cs:32]: I would avoid including changes irrelevant from the PR's target, but I guess this one may be fine.

File: osu.Game/Configuration/OsuConfigManager.cs (hunk 2 above)
Comment (peppy) [reply osu.Game/Configuration/OsuConfigManager.cs:418]: `TouchDisableGameplayTaps`

File: osu.Game/Overlays/Settings/Sections/Input/TouchSettings.cs (hunk 3 above)
Comment (peppy) [reply osu.Game/Overlays/Settings/Sections/Input/TouchSettings.cs:32]: I'd still prefer it to be separate, for better changelog tracking.

File: osu.Game/Rulesets/UI/RulesetInputManager.cs (hunk 1 above)
Comment (peppy) [reply osu.Game/Rulesets/UI/RulesetInputManager.cs:156]: > This path is specific to the osu! ruleset

How so? I don't understand. This should apply to any ruleset which supports mouse but not explicitly touch?

File: osu.Game/Screens/Play/PlayerSettings/InputSettings.cs [hunk 4]
```diff
@@ -19,12 +20,13 @@ public InputSettings()
             {
                 mouseButtonsCheckbox = new PlayerCheckbox
```
Comment (peppy) [inline osu.Game/Screens/Play/PlayerSettings/InputSettings.cs:21]: I'd rename the variable to `disableGameplayInputCheckbox`.

File: osu.Game/Overlays/Settings/Sections/Input/TouchSettings.cs (hunk 3 above)
Comment (Susko3) [reply osu.Game/Overlays/Settings/Sections/Input/TouchSettings.cs:32]: I can move this and the `OsuGameAndroid` change to a new pull titled "Show touch settings on Android" if you wish.

File: osu.Game/Rulesets/UI/RulesetInputManager.cs (hunk 1 above)
Comment (frenzibyte) [reply osu.Game/Rulesets/UI/RulesetInputManager.cs:156]: Hmm, I guess freeform custom rulesets would end up with a non-functioning setting that way...fair enough.

File: osu.Game/Screens/Play/PlayerSettings/InputSettings.cs (hunk 4 above)
Comment (Susko3) [reply osu.Game/Screens/Play/PlayerSettings/InputSettings.cs:21]: 👍 
Missed that one in the refactor. But I think I'll just move this to BDL and scrap the variable entirely.

File: osu.Game/Overlays/Settings/Sections/Input/TouchSettings.cs (hunk 3 above)
Comment (peppy) [reply osu.Game/Overlays/Settings/Sections/Input/TouchSettings.cs:32]: Yep, sounds good.

File: osu.Game/Screens/Play/PlayerSettings/InputSettings.cs [hunk 5]
```diff
@@ -23,9 +23,8 @@ private void load(OsuConfigManager config)
             {
                 new PlayerCheckbox
                 {
-                    // TODO: change to touchscreen detection once https://github.com/ppy/osu/pull/25348 makes it in
```
Comment (peppy) [inline osu.Game/Screens/Play/PlayerSettings/InputSettings.cs]: I felt like this change made sense, but I guess it's not a big deal either way for now.

**Review by peppy — APPROVED** [review]

**Pull request merged**